igvc_perception/src/projection/main.cpp

```diff
@@ -0,0 +1,276 @@
+#include <cv_bridge/cv_bridge.h>
+#include <image_geometry/pinhole_camera_model.h>
+#include <pcl/kdtree/kdtree_flann.h>
+#include <pcl/point_cloud.h>
+#include <pcl/point_types.h>
+#include <pcl_ros/point_cloud.h>
+#include <ros/publisher.h>
+#include <ros/ros.h>
+#include <sensor_msgs/Image.h>
+#include <tf/transform_listener.h>
+#include <igvc_utils/NodeUtils.hpp>
+#include "velodyne_pointcloud/point_types.h"
+
+/**
+ * Reads from velodyne pointcloud and uses a pinhole camera model in order to correlate pixels to 3d points.
+ * Publishes to a pointcloud topic, line_pointcloud, containing all 3d points that are part of a line.
+ */
+
+image_geometry::PinholeCameraModel cam_model;
+tf::StampedTransform transform_lidar_to_cam;
+tf::StampedTransform transform_cam_to_base;
+tf::StampedTransform transform_lidar_to_base;
+ros::Publisher _line_pointcloud_pub;
+ros::Subscriber image_sub;
+ros::Subscriber lidar_pointcloud_sub;
+sensor_msgs::CameraInfo::ConstPtr cam_info;
+cv::Mat cv_img;
+int resize_width;
+int resize_height;
+std::string line_topic;
+std::string pointcloud_topic;
+std::string lidar_topic;
+
+/**
+ * Finds the two nearest neighbors to the input point, constraining the nearest neighbors
+ * to be from different lidar channels.
+ *
+ * @param point The point to find the nearest neighbors of
+ * @param kdtree_list List of 16 kdtrees, one for each lidar channel
+ * @param nearest1 The out parameter for the nearest pixel
+ * @param nearest2 the out parameter for the second nearest pixel from a different channel
+ */
+
+void find_nearest_neighbors(pcl::PointXY point, const std::vector<pcl::KdTreeFLANN<pcl::PointXY>> &kdtree_list,
+                            pcl::PointXY &nearest1, pcl::PointXY &nearest2)
+{
+  float nearest1Dist = FLT_MAX;
```

Comment:
use snake_case for variables according to ros c++ style guide.

http://wiki.ros.org/CppStyleGuide#Variables